src/coreclr/jit/fgprofile.cpp

```diff
@@ -439,443 +447,1176 @@ void BlockCountInstrumentor::InstrumentMethodEntry(Schema& schema, BYTE* profile
+                        {
+                            continue;
+                        }
 
-        JITDUMP("Modified call is now\n");
-        DISPTREE(call);
+                        BlockSetOps::AddElemD(comp, processed, i);
 
-        // Restore the stub address on the call
-        //
-        call->gtStubCallStubAddr = call->gtClassProfileCandidateInfo->stubAddr;
+                        if (BlockSetOps::IsMember(comp, marked, target->bbNum))
+                        {
+                            visitor->VisitNonTreeEdge(block, target, SpanningTreeVisitor::Kind::Target);
+                        }
+                        else
+                        {
+                            visitor->VisitTreeEdge(block, target);
+                            stack.Push(target);
+                            BlockSetOps::AddElemD(comp, marked, target->bbNum);
+                        }
+                    }
 
-        m_instrCount++;
+                    // Critical edges
+                    //
+                    for (unsigned i = 0; i < numSucc; i++)
+                    {
+                        BasicBlock* const target = scratch.Top(i);
+
+                        if (BlockSetOps::IsMember(comp, processed, i))
+                        {
+                            continue;
+                        }
+
+                        BlockSetOps::AddElemD(comp, processed, i);
+
+                        if (BlockSetOps::IsMember(comp, marked, target->bbNum))
+                        {
+                            visitor->VisitNonTreeEdge(block, target, SpanningTreeVisitor::Kind::Edge);
+                        }
+                        else
+                        {
+                            visitor->VisitTreeEdge(block, target);
+                            stack.Push(target);
+                            BlockSetOps::AddElemD(comp, marked, target->bbNum);
+                        }
+                    }
+
+                    // Verify we processed each successor.
+                    //
+                    assert(numSucc == BlockSetOps::Count(comp, processed));
+                }
+            }
+            break;
+        }
     }
-};
+}
 
 //------------------------------------------------------------------------
-// SuppressProbesFunctor: functor that resets IR back to the state
-//   it had if there were no class probes.
+// EfficientEdgeCountInstrumentor: instrumentor that adds a counter to
+//   selective edges.
 //
-class SuppressProbesFunctor
+// Based on "Optimally Profiling and Tracing Programs,"
+// Ball and Larus PLDI '92.
+//
+class EfficientEdgeCountInstrumentor : public Instrumentor, public SpanningTreeVisitor
 {
 private:
-    unsigned& m_cleanupCount;
+    // A particular edge probe. These are linked
+    // on the source block via bbProbeList.
+    //
+    struct Probe
+    {
+        BasicBlock* target;
+        Probe*      next;
+        int         schemaIndex;
+        Kind        kind;
```

Comment:
IMO, `Kind` is too generic of a name

Reply:
Changed to `EdgeKind` and made the enumerants more descriptive.